### PR TITLE
cluster/dm: support backup and restore cluster meta

### LINKED ISSUE #1782: support backup  meta.yaml
## Feature Request

**Is your feature request related to a problem? Please describe:**

If user deleted the  meta.yaml file in TiUP server, it is hard to recover it. We need to provide an easy way to backup it. 

**Describe the feature you'd like:**

backup command :  tiup-cluster meta backup 
restore command :  tiup-cluster meta restore

**Why the featue is needed:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by nexustar**
### What problem does this PR solve? 

- close #1782

### What is changed and how it works?

Screenshot from 2022-03-18 15-39-19

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

Release notes:

```release-note
NONE
```

### CODE REVIEW

File: components/dm/command/meta.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	cmd := &cobra.Command{
+		Use:   "meta",
+		Short: "backup/restore meta information",
+	}
+
+	var filePath string
+
+	var metaBackupCmd = &cobra.Command{
+		Use:   "backup <cluster-name>",
+		Short: "backup topology and other information of cluster",
```
Comment (AstroProfundis) [inline components/dm/command/meta.go:33]:
```suggestion
		Short: "backup topology and other information of a cluster",
```

File: components/dm/command/meta.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 1 {
+				return fmt.Errorf("please input cluster-name")
```
Comment (AstroProfundis) [inline components/dm/command/meta.go:36]:
```suggestion
				return fmt.Errorf("please input cluster name")
```

File: components/dm/command/meta.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+				log.Infof("successfully backup meta of cluster %s on %s", args[0], filePath)
+			}
+			return err
+		},
+	}
+	metaBackupCmd.Flags().StringVar(&filePath, "file", "", "filepath of output tarball")
+
+	var metaRestoreCmd = &cobra.Command{
+		Use:   "restore <cluster-name> <backup-file>",
+		Short: "restore topology and other information of cluster",
```
Comment (AstroProfundis) [inline components/dm/command/meta.go:52]:
```suggestion
		Short: "restore topology and other information of a cluster",
```

File: components/dm/command/meta.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 2 {
+				return fmt.Errorf("please input cluster-name and backup-file")
```
Comment (AstroProfundis) [inline components/dm/command/meta.go:55]:
```suggestion
				return fmt.Errorf("please input cluster name and path to the backup file")
```

File: pkg/cluster/manager/basic.go [hunk 5]
```diff
@@ -302,3 +304,49 @@ func checkTiFlashWithTLS(topo spec.Topology, version string) error {
...
+	}
+	err = os.RemoveAll(m.specManager.Path(clusterName))
+	if err != nil {
+		return err
+	}
+	f, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	return utils.Untar(f, m.specManager.Path(clusterName))
```
Comment (AstroProfundis) [inline pkg/cluster/manager/basic.go:351]: I think we may need to check for cross conflicts of directories like we do in `deploy` here.

**Review by srstack — APPROVED** [review]

**Comment by srstack** [comment]
remove LGTM

**Review by AstroProfundis — APPROVED** [review]

**Pull request merged**